strutil/quantity/quantity.go

```diff
@@ -200,3 +201,262 @@ func FormatDuration(dt float64) string {
+	if dt <= 0 {
+		return "0" + units[min]
+	}
+
+	// Special case: render as "ages!"
+	if dt >= float64(math.MaxUint64) {
+		return "ages!"
+	}
+
+	// Fractional seconds to nanoseconds
+	delta := Duration(dt * float64(Second))
+
+	// If we only render a subset of place values, we need to apply
+	// the render mode to the last place value. We do this indirectly
+	// by moving up the 'min' accuracy.
+	if count == ShowCompact {
+		for i, key := range keys {
+
+			// The duration is smaller than this place value, skip
+			if delta < key {
+				continue
+			}
+
+			// Compact mode only renders two places, up until seconds
+			// The indexing is safe as places beyond seconds exist
+			if key >= Minute {
+				index := i + 1
+				if keys[index] > min {
+					min = keys[index]
+				}
+				break
+			}
+
+			// Only one place if we are s, ms, us or ns
+			if key < Minute {
+				if key > min {
+					min = key
+				}
+				break
+			}
+		}
+	}
+
+	// Apply the rendermode to the least significant place value
+	mod := delta % min
+	switch mode {
+	case TimeLeft:
+		// Ceiling
+		if mod > 0 {
+			delta += min - mod
+		}
+	case TimePassed:
+		// Floor
+		delta -= mod
+	case TimeRounded:
+		if mod >= (min / 2) {
+			// Rounded Up
+			delta += min - mod
+		} else {
+			// Rounded Down
+			delta -= mod
+		}
+	default:
+		// Invalid mode
+		return "inv!"
+	}
+
+	// Special case: less than minimum accuracy
+	if delta < min {
+		return "0" + units[min]
+	}
+
+	// In ShowCompact mode only 2-digit days are possible
+	if count == ShowCompact && delta >= (100*Day) {
+		return "ages!"
+	}
+
+	// We now take the delta nanoseconds and render it in terms
+	// of readable units. We subtract the rendered part as we
+	// iterate through the place values (y, d, h, m, s, ...)
+	remainingDelta := delta
+
+	// Iterate through units of time in order from
+	// highest to lowest (years, days, ...)
+	for _, key := range keys {
+		// Nothing left to render, do not bother iterating any further
+		if remainingDelta <= 0 {
+			break
+		}
+
+		// No place value left greater than required
+		// accuracy to render.
+		if remainingDelta < min && len(render) != 0 {
+			break
+		}
+
+		// Remainder is less than current place value,
+		// skip to next.
+		if remainingDelta < key {
+			continue
+		}
+
+		// If the duration exceeds the maximum place value
+		// (unit if time) we will not render it, just return
+		// "ages!".
+		if key > max {
+			return "ages!"
+		}
+
+		unit := remainingDelta / key
+		remainder := remainingDelta % key
+
+		// If the accuracy is less than 1s, we support
+		// rendering a fractional second part.
+		if remainingDelta < Minute && remainder > 0 && count == ShowVerbose {
+			digits := 0
+
+			// Number of digits required to render
+			// nano second fractunal accuracy.
+			for cc := key; cc > 1; cc /= 10 {
+				digits++
+			}
+
+			// Number of digits to render specified
+			// fractunal accuracy.
+			for cc := min; cc > 1; cc /= 10 {
+				digits--
```

Comment:
these two loops are hard to understand (at least to me), unless some comment is added to explain why we are subtracting them: maybe rewriting them with different variables would make it more obvious:

```
digitsForNanoSeconds = ...  // first for loop
digitsForMinAccuracy = ... // second for loop, with ++
digits := digitsForNanoSeconds - digitsForMinAccuracy
```
though, honestly, even like this I don't understand it :-) It's not clear to me why we just don't use digitsForMinAccuracy, and why we don't take into account the `count` variable that tells us how many spaces we can use.

Reply:
The short reason is that places are counted from the right to left, in other words, starting at the nano-second place value.

I have reworked and simplified it with much better explanations. Let me know what you think!